feat: add custom transaction schema to formatTransaction

nicolasbrugneaux opened the pull request:
Greetings 👋 


## Description

This PR aims to solidify the Plugin experience by allowing a `customTransactionSchema` to be used by `formatTransaction`. This seems to have be thought of before as the `formatTransaction` already accepted a custom schema in its parameters but it seems it wasn't used anywhere yet. This PR aims solve that.

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run lint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

nicolasbrugneaux committed:
feat: add custom transaction schema to formatTransaction

nicolasbrugneaux committed:
docs: add entries to changelogs

nicolasbrugneaux committed:
tests: initial tests

jdevcs reviewed (commented):
@nicolasbrugneaux Thanks for your contribution, there are some build issues that need fix.

nicolasbrugneaux committed:
fix: unused vars

nicolasbrugneaux committed:
tests: lint issues

nicolasbrugneaux commented:
> @nicolasbrugneaux Thanks for your contribution, there are some build issues that need fix.

I'm having issues running it locally but I think I fixed it all. Any chance we can re-run ci?

jdevcs commented:
some of tests needs update, including unit test:
```
web3-eth:  FAIL  test/unit/format_transaction.test.ts
web3-eth:   ● Test suite failed to run
web3-eth:     test/unit/format_transaction.test.ts:123:52 - error TS2339: Property 'feeCurrency' does not exist on type '{ from?: string | undefined; to?: string | null | undefined; value?: bigint | undefined; accessList?: { readonly address?: string | undefined; readonly storageKeys?: string[] | undefined; }[] | undefined; ... 17 more ...; s?: string | undefined; }'.
web3-eth:     123   expect(formatTransaction(customFieldTransaction).feeCurrency).toBeUndefined();
web3-eth:                                                            ~~~~~~~~~~~
web3-eth:     test/unit/format_transaction.test.ts:130:7 - error TS2322: Type '{ properties: { feeCurrency: string; from: { format: string; }; to: { oneOf: ({ format: string; type?: undefined; } | { type: string; format?: undefined; })[]; }; value: { format: string; }; gas: { format: string; }; ... 17 more ...; s: { ...; }; }; type: string; }' is not assignable to type 'ValidationSchemaInput | { type: string; properties: { from: { format: string; }; to: { oneOf: ({ format: string; type?: undefined; } | { type: string; format?: undefined; })[]; }; ... 19 more ...; s: { ...; }; }; } | undefined'.
web3-eth:       Types of property 'properties' are incompatible.
web3-eth:         Type '{ feeCurrency: string; from: { format: string; }; to: { oneOf: ({ format: string; type?: undefined; } | { type: string; format?: undefined; })[]; }; value: { format: string; }; gas: { format: string; }; gasPrice: { format: string; }; ... 16 more ...; s: { ...; }; }' is not assignable to type '{ from: { format: string; }; to: { oneOf: ({ format: string; type?: undefined; } | { type: string; format?: undefined; })[]; }; value: { format: string; }; gas: { format: string; }; gasPrice: { format: string; }; effectiveGasPrice: { ...; }; ... 15 more ...; s: { ...; }; }'.
web3-eth:           Object literal may only specify known properties, and 'feeCurrency' does not exist in type '{ from: { format: string; }; to: { oneOf: ({ format: string; type?: undefined; } | { type: string; format?: undefined; })[]; }; value: { format: string; }; gas: { format: string; }; gasPrice: { format: string; }; effectiveGasPrice: { ...; }; ... 15 more ...; s: { ...; }; }'.
web3-eth:     130       feeCurrency: 'address',
web3-eth:               ~~~~~~~~~~~~~~~~~~~~~~
web3-eth:     test/unit/format_transaction.test.ts:133:7 - error TS2339: Property 'feeCurrency' does not exist on type '{ from?: string | undefined; to?: string | null | undefined; value?: bigint | undefined; accessList?: { readonly address?: string | undefined; readonly storageKeys?: string[] | undefined; }[] | undefined; ... 17 more ...; s?: string | undefined; }'.
web3-eth:     133    }).feeCurrency,
web3-eth:               ~~~~~~~~
```

avkos commented on packages/web3-eth/CHANGELOG.md:
```diff
@@ -267,6 +267,14 @@ Documentation:
 
 -   Adds transaction property to be an empty list rather than undefined when no transactions are included in the block (#7151)
--   Change method `getTransactionReceipt` to not be casted as `TransactionReceipt` to give proper return type (#7159)
+    w
```
I think it was deleted by mistake

nicolasbrugneaux replied:
Oops, indeed.

avkos commented on packages/web3-eth/src/rpc_method_wrappers.ts:
```diff
@@ -296,10 +296,10 @@ export async function getBlock<ReturnFormat extends DataFormat>(
 			...res,
 			transactions: res.transactions ?? [],
-		}
+		};
```
Please try to not format existing code

nicolasbrugneaux replied:
Gotcha. format-on-save guilty once again :)

avkos commented on packages/web3-eth/src/utils/decode_signed_transaction.ts:
```diff
@@ -35,7 +36,10 @@ import { formatTransaction } from './format_transaction.js';
-	options: { fillInputAndData?: boolean } = { fillInputAndData: false },
+	options: { fillInputAndData?: boolean; transactionSchema?: ValidationSchemaInput } = {
+		fillInputAndData: false,
+		transactionSchema: undefined,
```
It can be optional so you can do not set it as `undefined`

nicolasbrugneaux committed:
refactor: pr review

nicolasbrugneaux committed:
fix: type errors

nicolasbrugneaux committed:
fix: dependency cycle

nicolasbrugneaux committed:
Merge branch '4.x' into nicolasbrugneaux/customTransactionSchema

avkos reviewed (changes requested):
LGTM. But still, some tests fail. Will be great to fix them cause I see some types incompatible.

Muhammad-Altabba reviewed (commented):
Thanks @nicolasbrugneaux,
Could you please revert the code formatting done in this MR so we can easily review it? (We are working on a better formatting for the code which will be in a separate MR).
Also please fix the tests so we can look at your MR deeply.
Thanks,

nicolasbrugneaux committed:
refactor: revert whitespaces changes

nicolasbrugneaux committed:
fix: types

nicolasbrugneaux committed:
Merge branch '4.x' into nicolasbrugneaux/customTransactionSchema

nicolasbrugneaux commented:
> Thanks @nicolasbrugneaux, Could you please revert the code formatting done in this MR so we can easily review it? (We are working on a better formatting for the code which will be in a separate MR). Also please fix the tests so we can look at your MR deeply. Thanks,

I tried fixing all whitespaces changes I could find, the ones left are the ones were I added more parameters to the method call so I wasn't sure what the right format should be.  🙏  Shouldn't impact readability too much IMHO.

I'm also confused why would any of my changes have an impact on the e2e tests failing, now that I (hopefully) fixed the type for node 18 too.

nicolasbrugneaux committed:
fix: types

aaronmgdr commented on packages/web3-eth-personal/src/rpc_method_wrappers.ts:
```diff
@@ -83,7 +85,9 @@ export const signTransaction = async (
 	passphrase: string,
 ) => {
-	const formattedTx = formatTransaction(tx, ETH_DATA_FORMAT);
+	const formattedTx = formatTransaction(tx, ETH_DATA_FORMAT, {
+		transactionSchema: requestManager.config.customTransactionSchema,
+	});
```
i might be off here but the test failing on node 18 test/unit/eth_personal.test.ts  might pass if the third arg here is only passed in when `requestManager.config.customTransactionSchema` is truthy

nicolasbrugneaux committed:
test: fix web3-eth-personal tests

nicolasbrugneaux commented:
Last failing tests should be now green 🙏

luu-alex commented:
@nicolasbrugneaux  thanks for your contribution, the tests look like they are passing! @jdevcs  if everything looks good we can merge this, and i can include this for release next week.

jdevcs commented on packages/web3-core/src/web3_batch_request.ts:
```diff
@@ -18,6 +18,7 @@ along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
 import { JsonRpcBatchResponse, JsonRpcOptionalRequest, JsonRpcRequest } from 'web3-types';
 import { jsonRpc, Web3DeferredPromise } from 'web3-utils';
 import { OperationAbortError, OperationTimeoutError, ResponseError } from 'web3-errors';
+// eslint-disable-next-line import/no-cycle
```
why eslint-disable-next-line import/no-cycle line is added?

jdevcs commented on packages/web3-core/src/web3_context.ts:
```diff
@@ -36,6 +36,7 @@ import { ExtensionObject, RequestManagerMiddleware } from './types.js';
 import { Web3BatchRequest } from './web3_batch_request.js';
 // eslint-disable-next-line import/no-cycle
 import { Web3Config, Web3ConfigEvent, Web3ConfigOptions } from './web3_config.js';
+// eslint-disable-next-line import/no-cycle
```
is web3_request_manager importing from context?

jdevcs commented on packages/web3-core/src/web3_request_manager.ts:
```diff
@@ -74,15 +75,19 @@ export class Web3RequestManager<
 }> {
 	private _provider?: SupportedProviders<API>;
 	private readonly useRpcCallSpecification?: boolean;
+	public config: Partial<Web3ConfigOptions>;
```
web3 config is already always in web3-context so it should be used from there in other parts of lib, so considering that it should not be added on another level ( request manager ) https://github.com/web3/web3.js/blob/b3cb1b782cc2c62bd87909c7cdc426ed43a49c1c/packages/web3-core/src/web3_context.ts#L51

jdevcs commented on packages/web3-eth-personal/src/rpc_method_wrappers.ts:
```diff
@@ -73,7 +73,9 @@ export const sendTransaction = async (
 	passphrase: string,
 ) => {
-	const formattedTx = formatTransaction(tx, ETH_DATA_FORMAT);
+	const formattedTx = formatTransaction(tx, ETH_DATA_FORMAT, {
```
I suggested in above comments in request manager to lets remove config from request manager so with that , this file also needs change, and in `Personal` class use context, to get tx formatting schema from context and pass to these functions, and it should be optional so its not breaking change.

jdevcs commented:
@nicolasbrugneaux great effort, for this feature. I suggested few changes. Thanks for your contributions

nicolasbrugneaux committed:
refactor: remove config from manager

nicolasbrugneaux commented:
@jdevcs thanks for the thourough review. I should have addressed all feedback. Let me know if you were thinking of doing it differently. 

Side note: I also refactored the type of `config.customTransactionSchema` to be more generic and thus not have to `import {CustomTransactionSchema} from 'web3-eth'` into `web3-core`. Let me know if you think there's a better place for that new type.

nicolasbrugneaux committed:
Merge branch '4.x' into nicolasbrugneaux/customTransactionSchema

nicolasbrugneaux committed:
fix: types

luu-alex commented:
Hey @nicolasbrugneaux , https://github.com/web3/web3.js/actions/runs/10886942923/job/30214851919?pr=7227 the build is currently failing

nicolasbrugneaux committed:
fix: build issue

nicolasbrugneaux commented:
Fixed the build, builds locally at least 😅

luu-alex commented:
almost there 😅  
web3-core:  FAIL  test/unit/web3_context.test.ts
web3-core:   ● Test suite failed to run
web3-core:     test/unit/web3_context.test.ts:127:37 - error TS2339: Property 'config' does not exist on type 'Web3RequestManager<unknown>'.
web3-core:     127    expect(newContext.requestManager.config.defaultHardfork).toEqual(
web3-core:

nicolasbrugneaux committed:
fix: CustomTransactionSchema type

nicolasbrugneaux commented:
I noticed the interface for `config.customTransactionSchema` was incorrect so I fixed that as well as the failing test above.

nicolasbrugneaux commented:
Thank you!!

nicolasbrugneaux commented:
hey there @luu-alex since you kindly merged this I wanted to bring your attention to a follow-up issue I encountered: 
- https://github.com/web3/web3.js/issues/7273
